standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandlerContext.java

```diff
@@ -0,0 +1,173 @@
+  private HMSHandler hmsHandler;
+
+  // Thread local configuration is needed as many threads could make changes
+  // to the conf using the connection hook
+  private Configuration configuration;
+
+  // Thread local Map to keep track of modified meta conf keys
+  private Map<String, String> modifiedConfig = new HashMap<>();
+
+  private Integer threadId = nextSerialNum.incrementAndGet();
+  // This will only be set if the metastore is being accessed from a metastore Thrift server,
+  // not if it is from the CLI. Also, only if the TTransport being used to connect is an
+  // instance of TSocket. This is also not set when kerberos is used.
+  private String ipAddress;
+
+  private Map<String, com.codahale.metrics.Timer.Context> timerContexts = new HashMap<>();
+
+  private HMSHandlerContext() {
+
+  }
+
+  public static Optional<RawStore> getRawStore() {
+    HMSHandlerContext ctx = context.get();
+    return ctx != null ? ctx.getLocalRawStore() : Optional.empty();
+  }
+
+  public static Optional<HMSHandler> getHMSHandler() {
+    HMSHandlerContext ctx = context.get();
+    return ctx != null ? ctx.getLocalHmsHandler() : Optional.empty();
+  }
+
+  public static Optional<String> getIpAddress() {
+    HMSHandlerContext ctx = context.get();
+    return ctx != null ? ctx.getRemoteIpAddress() : Optional.empty();
+  }
+
+  public static Optional<Configuration> getConfiguration() {
+    HMSHandlerContext ctx = context.get();
+    return ctx != null ? ctx.getLocalConfiguration() : Optional.empty();
+  }
+
+  public static TxnStore getTxnStore(Configuration conf) {
+    return getContext().getLocalTxnStore().orElseGet(() -> {
+      TxnStore txnStore = TxnUtils.getTxnStore(conf);
+      setTxnStore(txnStore);
+      return txnStore;
+    });
+  }
+
+  public static Map<String, String> getModifiedConfig() {
+    return getContext().modifiedConfig;
+  }
+
+  public static Integer getThreadId() {
+    return getContext().threadId;
+  }
+
+  public static Map<String, com.codahale.metrics.Timer.Context> getTimerContexts() {
+    return getContext().timerContexts;
+  }
+
+  private static HMSHandlerContext getContext() {
+    HMSHandlerContext ctx = context.get();
+    if (ctx == null) {
+      context.set(ctx = new HMSHandlerContext());
+    }
+    return ctx;
+  }
+
+  public static void setRawStore(RawStore rawStore) {
+    getContext().rawStore = rawStore;
+  }
+
+  public static void setTxnStore(TxnStore txnStore) {
+    getContext().txnStore = txnStore;
+  }
+
+  public static void setHMSHandler(HMSHandler hmsHandler) {
+    getContext().hmsHandler = hmsHandler;
+  }
+
+  public static void setConfiguration(Configuration conf) {
+    getContext().configuration = conf;
+  }
+
+  public static void setIpAddress(String ipAddress) {
+    getContext().ipAddress = ipAddress;
+  }
+
+  public static void clear(CleanupHook cleanupHook) {
+    HMSHandlerContext ctx = context.get();
+    context.remove();
+    if (ctx != null && cleanupHook != null) {
```

Comment:
Feels a bit like overengineering to me.
Why not just do this the `cleanupHook` stuff outside, and clear the values here?

Reply:
OK, let me take a look. Thanks